ase/sw/protocol_backend.c

```diff
@@ -262,59 +340,109 @@ void wr_memline_dex(cci_pkt *pkt)
-	/*       ase_memcpy((char*)pkt->qword, rd_target_vaddr, CL_BYTE_WIDTH); */
-
-	/*       // Get cl_addr, deduct wr_target, use qw_start to determine exact qword */
-	/*       wr_target_vaddr = (uint64_t*)( (uint64_t)rd_target_vaddr + pkt->qw_start*8 ); */
-
-	/*       // CmpXchg output */
-	/*       pkt->success = (int)__sync_bool_compare_and_swap (wr_target_vaddr, cmp_qword, new_qword); */
-
-	/*       // Debug output */
-	/* #ifdef ASE_DEBUG */
-	/*        */
-	/*       ASE_DBG("CmpXchg_op=%d\n", pkt->success); */
-	/*        */
-	/* #endif */
-	/*     } */
-	/* #endif */
 
 	FUNC_CALL_EXIT;
 }
 
 
 /*
- * DPI: ReadLine Data exchange
+ * DPI: Write line response (from application)
+ */
+void wr_memline_rsp_dex(cci_pkt *pkt)
+{
+	FUNC_CALL_ENTRY;
+
+	ase_host_memory_write_rsp wr_rsp;
+	int status;
+	static int err_cnt;
+
+	// The only task required here is to consume the response from the application.
+	// triggered by wr_memline_req_dex. The response indicates whether the address
+	// was valid.  Raise an error for invalid addresses.
+
+	if (pkt->mode == CCIPKT_WRITE_MODE) {
+		while (true) {
+			status = mqueue_recv(app2sim_membus_wr_rsp_rx, (char *) &wr_rsp, sizeof(wr_rsp));
+
+			if (status == ASE_MSG_PRESENT) {
+				if (wr_rsp.status != HOST_MEM_STATUS_VALID) {
+					memline_addr_error("WRITE", wr_rsp.status, wr_rsp.pa, wr_rsp.va);
+					pkt->success = 0;
+				}
+				break;
+			}
+
+			// Error?  Probably channel closed and the simulator will be closing
+			// soon.  Allow a few errors but give up if it keeps happening.
+			if (status == ASE_MSG_ERROR) {
+				if (err_cnt++ > 10) {
+					ASE_ERR("Error receiving memory write line response\n");
+					start_simkill_countdown();
+				}
+				break;
+			}
+		}
+	}
+
+	FUNC_CALL_EXIT;
+}
+
+
+/*
+ * DPI: Read line request
  */
-void rd_memline_dex(cci_pkt *pkt)
+void rd_memline_req_dex(cci_pkt *pkt)
 {
 	FUNC_CALL_ENTRY;
 
 	uint64_t phys_addr;
-	uint64_t *rd_target_vaddr = (uint64_t *) NULL;
+	ase_host_memory_read_req rd_req;
 
-	// Get cl_addr, deduce rd_target_vaddr
 	phys_addr = (uint64_t) pkt->cl_addr << 6;
-	rd_target_vaddr = ase_fakeaddr_to_vaddr((uint64_t) phys_addr);
 
-	// Read from memory
-	ase_memcpy((char *) pkt->qword, rd_target_vaddr, CL_BYTE_WIDTH);
+	rd_req.req = HOST_MEM_REQ_READ_LINE;
+	rd_req.addr = phys_addr;
+
+	mqueue_send(sim2app_membus_rd_req_tx, (char *) &rd_req, sizeof(rd_req));
+
+	FUNC_CALL_EXIT;
+}
+
+
+/*
+ * DPI: Read line data response
+ */
+void rd_memline_rsp_dex(cci_pkt *pkt)
+{
+	FUNC_CALL_ENTRY;
+
+	ase_host_memory_read_rsp rd_rsp;
+	int status;
+	static int err_cnt;
+
+	while (true) {
+		status = mqueue_recv(app2sim_membus_rd_rsp_rx, (char *) &rd_rsp, sizeof(rd_rsp));
+
+		if (status == ASE_MSG_PRESENT) {
+			if (rd_rsp.status != HOST_MEM_STATUS_VALID) {
+				memline_addr_error("READ", rd_rsp.status, rd_rsp.pa, rd_rsp.va);
+			}
+
+			// Read from memory
+			ase_memcpy((char *) pkt->qword, rd_rsp.data, CL_BYTE_WIDTH);
```

Comment:
Yes, and this goes all the way down to the Verilog part of the simulator. I don't think it has much impact on simulation speed since each line is necessarily delivered in separate simulated cycles. The version I'm replacing was logically similar -- one call to the DPI read function per line, even in multi-line requests.

Reply:
Ok. That clarifies.